File: bin/eval_levelso1o.jl
using CSV, DataFrames, MixedModels, StatsBase, CategoricalArrays, JSON

if length(ARGS) < 2
    error("Usage: julia eval_levelso1o.jl <experiment_name> <backbone_name>")
end

experiment_name = ARGS[1]
backbone_name = ARGS[2]
csv_file = "data/evals/levelso1o/$(experiment_name)_$(backbone_name)_stats.csv"

if !isfile(csv_file)
    error("Stats file not found: $csv_file")
end

data = CSV.read(csv_file, DataFrame)

# Reshape from wide to long format
data_long = DataFrame()
for choice_idx in 0:2
    tmp = select(data, :participant_id, :y)
    tmp.choice_idx = fill(choice_idx, nrow(data))
    tmp.is_choice = Int.(data.y .== choice_idx)
    tmp.base_sim = data[:, Symbol("base_sim_$choice_idx")]
    tmp.ablation_sim = data[:, Symbol("metalign_sim_$choice_idx")]
    tmp.triplet_id = 1:nrow(data)
    append!(data_long, tmp)
end

data_long.participant_id = categorical(data_long.participant_id)
data_long.triplet_id = categorical(data_long.triplet_id)

# Determine if this is main or an ablation
is_main = experiment_name == "MAIN"

# If ablation, load the main model data to get full metalign
if !is_main
    main_csv_file = "data/evals/levelso1o/MAIN_$(backbone_name)_stats.csv"
    if !isfile(main_csv_file)
        error("Main model stats file not found: $main_csv_file")
    end
    main_data = CSV.read(main_csv_file, DataFrame)
    
    # Add main metalign similarities to long format data
    main_long = DataFrame()
    for choice_idx in 0:2
        tmp = DataFrame(
            triplet_id = 1:nrow(main_data),
            choice_idx = fill(choice_idx, nrow(main_data)),
            main_sim = main_data[:, Symbol("metalign_sim_$choice_idx")]
        )
        append!(main_long, tmp)
    end
    
    # Merge with ablation data
    data_long = leftjoin(data_long, main_long, on=[:triplet_id, :choice_idx])
end

# Normalize predictors to help convergence
dropmissing!(data_long)
data_long.base_sim = zscore(data_long.base_sim)
data_long.ablation_sim = zscore(data_long.ablation_sim)
if "main_sim" in names(data_long)
    data_long.main_sim = zscore(data_long.main_sim)
end

if is_main
    # Model 0: Base only
    formula_0 = @formula(is_choice ~ base_sim + (1 + base_sim | participant_id))
    model_0 = fit(MixedModel, formula_0, data_long, Bernoulli())
    
    println("=" ^ 80)
    println("MODEL 0: Base only")
    println("=" ^ 80)
    println(model_0)
    println()
    # Main model: only compare M0 vs M1
    formula_1 = @formula(is_choice ~ base_sim + ablation_sim + (1 + base_sim + ablation_sim | participant_id))
    model_1 = fit(MixedModel, formula_1, data_long, Bernoulli())
    
    println("=" ^ 80)
    println("MODEL 1: Base + Full Metalign")
    println("=" ^ 80)
    println(model_1)
    println()
    
    lrt = MixedModels.likelihoodratiotest(model_0, model_1)
    println("=" ^ 80)
    println("LIKELIHOOD RATIO TEST")
    println("=" ^ 80)
    println(lrt)
    println()
    
    println("=" ^ 80)
    println("MODEL COMPARISON")
    println("=" ^ 80)
    println("Model 0 - AIC: ", round(aic(model_0), digits=2), " BIC: ", round(bic(model_0), digits=2))
    println("Model 1 - AIC: ", round(aic(model_1), digits=2), " BIC: ", round(bic(model_1), digits=2))
    println()
    
    delta_bic = bic(model_0) - bic(model_1)
    log10_bf = (delta_bic / 2) / log(10)
    
    println("=" ^ 80)
    println("BAYESIAN MODEL COMPARISON (BIC Approximation)")
    println("=" ^ 80)
    println("Delta BIC: ", round(delta_bic, digits=2))
    println("Log10(Bayes Factor B10): ~", round(log10_bf, digits=2))
    println()
    
    # Save results
    results = Dict(
        "experiment_name" => experiment_name,
        "backbone_name" => backbone_name,
        "is_main_model" => true,
        "model_0" => Dict(
            "aic" => aic(model_0),
            "bic" => bic(model_0),
            "loglikelihood" => loglikelihood(model_0),
            "fixed_effects" => Dict(
                "intercept" => coef(model_0)[1],
                "base_sim" => coef(model_0)[2]
            )
        ),
        "model_1" => Dict(
            "aic" => aic(model_1),
            "bic" => bic(model_1),
            "loglikelihood" => loglikelihood(model_1),
            "fixed_effects" => Dict(
                "intercept" => coef(model_1)[1],
                "base_sim" => coef(model_1)[2],
                "metalign_sim" => coef(model_1)[3]
            )
        ),
        "lrt_0_vs_1" => Dict(
            "deviance" => lrt.deviance[2],
            "dof" => lrt.dof[2],
            "pvalue" => lrt.pval[2]
        ),
        "comparison_0_vs_1" => Dict(
            "delta_aic" => aic(model_0) - aic(model_1),
            "delta_bic" => delta_bic,
            "log10_bayes_factor" => log10_bf
        )
    )
else
    # Ablation model: only compare M1 vs M2
    formula_1 = @formula(is_choice ~ base_sim + ablation_sim + (1 + base_sim + ablation_sim | participant_id))
    model_1 = fit(MixedModel, formula_1, data_long, Bernoulli())
    
    println("=" ^ 80)
    println("MODEL 1: Base + Ablation ($experiment_name)")
    println("=" ^ 80)
    println(model_1)
    println()
    
    formula_2 = @formula(is_choice ~ base_sim + ablation_sim + main_sim + (1 + base_sim + ablation_sim + main_sim | participant_id))
    model_2 = fit(MixedModel, formula_2, data_long, Bernoulli())
    
    println("=" ^ 80)
    println("MODEL 2: Base + Ablation + Full Metalign")
    println("=" ^ 80)
    println(model_2)
    println()
    
    lrt = MixedModels.likelihoodratiotest(model_1, model_2)
    
    println("=" ^ 80)
    println("LIKELIHOOD RATIO TEST: Model 1 vs Model 2")
    println("=" ^ 80)
    println(lrt)
    println()
    
    println("=" ^ 80)
    println("MODEL COMPARISON")
    println("=" ^ 80)
    println("Model 1 - AIC: ", round(aic(model_1), digits=2), " BIC: ", round(bic(model_1), digits=2))
    println("Model 2 - AIC: ", round(aic(model_2), digits=2), " BIC: ", round(bic(model_2), digits=2))
    println()
    
    delta_bic = bic(model_1) - bic(model_2)
    log10_bf = (delta_bic / 2) / log(10)
    
    println("=" ^ 80)
    println("BAYESIAN MODEL COMPARISON (BIC Approximation)")
    println("=" ^ 80)
    println("Delta BIC: ", round(delta_bic, digits=2), " | Log10(BF): ~", round(log10_bf, digits=2))
    println()
    
    # Save results
    results = Dict(
        "experiment_name" => experiment_name,
        "backbone_name" => backbone_name,
        "is_main_model" => false,
        "model_1" => Dict(
            "aic" => aic(model_1),
            "bic" => bic(model_1),
            "loglikelihood" => loglikelihood(model_1),
            "fixed_effects" => Dict(
                "intercept" => coef(model_1)[1],
                "base_sim" => coef(model_1)[2],
                "ablation_sim" => coef(model_1)[3]
            )
        ),
        "model_2" => Dict(
            "aic" => aic(model_2),
            "bic" => bic(model_2),
            "loglikelihood" => loglikelihood(model_2),
            "fixed_effects" => Dict(
                "intercept" => coef(model_2)[1],
                "base_sim" => coef(model_2)[2],
                "ablation_sim" => coef(model_2)[3],
                "main_sim" => coef(model_2)[4]
            )
        ),
        "lrt_1_vs_2" => Dict(
            "deviance" => lrt.deviance[2],
            "dof" => lrt.dof[2],
            "pvalue" => lrt.pval[2]
        ),
        "comparison_1_vs_2" => Dict(
            "delta_aic" => aic(model_1) - aic(model_2),
            "delta_bic" => delta_bic,
            "log10_bayes_factor" => log10_bf
        )
    )
end

# Save results to JSON file
output_file = "data/evals/levelso1o/$(experiment_name)_$(backbone_name)_analysis.json"
open(output_file, "w") do f
    JSON.print(f, results, 4)
end

println("=" ^ 80)
println("Results saved to: ", output_file)
println("=" ^ 80)

File: bin/eval_levelso1o.R
source("bin/utils.R")
library(mlogit, quietly = TRUE)
library(tidyverse, quietly = TRUE)

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) {
  stop("Usage: Rscript eval_levelso1o.R <experiment_name> <backbone_name>")
}

experiment_name <- args[1]
backbone_name <- args[2]

cat(sprintf("Running eval_levelso1o.R for %s %s\n", experiment_name, backbone_name))

stats_file <- sprintf("data/evals/levelso1o/%s_%s_stats.csv", experiment_name, backbone_name)
if (!file.exists(stats_file)) stop(sprintf("Stats file not found: %s", stats_file))

df <- read.csv(stats_file)
df <- df %>% mutate(trial_id = row_number())

is_main <- experiment_name == "MAIN"

if (!is_main) {
  main_stats_file <- sprintf("data/evals/levelso1o/MAIN_%s_stats.csv", backbone_name)
  if (!file.exists(main_stats_file)) stop("Main model stats file not found for ablation comparison")

  df_main <- read.csv(main_stats_file)

  df$metalign_sim_main_0 <- df_main$metalign_sim_0
  df$metalign_sim_main_1 <- df_main$metalign_sim_1
  df$metalign_sim_main_2 <- df_main$metalign_sim_2
}
# Columns of interest
cols_to_pivot <- c(
  "base_sim_0", "base_sim_1", "base_sim_2",
  "metalign_sim_0", "metalign_sim_1", "metalign_sim_2"
)
if (!is_main) {
  cols_to_pivot <- c(cols_to_pivot, "metalign_sim_main_0", "metalign_sim_main_1", "metalign_sim_main_2")
}

# Pivot long
df_long <- df %>%
  pivot_longer(
    cols = any_of(gsub("_[0-9]$", "", cols_to_pivot) %>% paste0("_", 0:2)), # Matches base_sim_0, etc.
    names_to = c(".value", "alt"),
    names_pattern = "(.*)_([0-9])$"
  ) %>%
  mutate(
    alt = as.numeric(alt),
    is_choice = (y == alt),
    participant_id = as.factor(participant_id),
    trial_id = as.factor(trial_id) # Unique trial ID
  )

# Explicit dummies for alternatives (Alternative Specific Constants)
df_long$asc1 <- as.numeric(df_long$alt == 1)
df_long$asc2 <- as.numeric(df_long$alt == 2)

# Z-score predictors
df_long$base_sim <- z_score(df_long$base_sim)
df_long$metalign_sim <- z_score(df_long$metalign_sim)
if (!is_main) {
  df_long$metalign_sim_main <- z_score(df_long$metalign_sim_main)
}

# mlogit data format
# id.var = participant_id to handle panel data (mixed logit)
# chid.var = trial_id (choice situation)
data_m <- mlogit.data(df_long,
  choice = "is_choice",
  shape = "long",
  alt.var = "alt",
  chid.var = "trial_id",
  id.var = "participant_id"
)

results <- list(
  experiment_name = experiment_name,
  backbone_name = backbone_name
)

if (is_main) {
  cat("Fitting Null Model (for R2)...\n")
  # Null model: only intercepts
  m_null <- mlogit(is_choice ~ asc1 + asc2 | 0,
    data = data_m,
    rpar = c(asc1 = "n", asc2 = "n"),
    panel = TRUE,
    iterlim = 2000
  )
  null_ll <- as.numeric(logLik(m_null))

  cat("Fitting Model 0 (Base only)...\n")


  # Model 0
  m0 <- mlogit(is_choice ~ base_sim + asc1 + asc2 | 0,
    data = data_m,
    rpar = c(base_sim = "n", asc1 = "n", asc2 = "n"),
    panel = TRUE,
    iterlim = 2000
  )

  stats0 <- calculate_model_stats(m0, null_model_ll = null_ll)
  cat(sprintf("Model 0 LogLik: %.2f, BIC: %.2f\n", stats0$loglik, stats0$bic))

  cat("Fitting Model 1 (Base + Metalign)...\n")
  m1 <- mlogit(is_choice ~ base_sim + metalign_sim + asc1 + asc2 | 0,
    data = data_m,
    rpar = c(base_sim = "n", metalign_sim = "n", asc1 = "n", asc2 = "n"),
    panel = TRUE,
    iterlim = 2000
  )

  stats1 <- calculate_model_stats(m1, null_model_ll = null_ll)
  cat(sprintf("Model 1 LogLik: %.2f, BIC: %.2f\n", stats1$loglik, stats1$bic))

  log10_bf <- calculate_log10_bf(stats0$bic, stats1$bic)
  lrt <- calculate_lrt(m0, m1)
  cat(sprintf("Log10 BF (M1/M0): %.2f, LRT p-value: %.3e\n", log10_bf, lrt$p_value))

  results$model_0 <- stats0
  results$model_1 <- stats1
  results$comparison_0_vs_1 <- list(
    log10_bayes_factor = log10_bf,
    lrt = lrt
  )
} else {
  # Ablation
  cat("Fitting Null Model (for R2)...\n")
  # Null model: only intercepts
  m_null <- mlogit(is_choice ~ asc1 + asc2 | 0,
    data = data_m,
    rpar = c(asc1 = "n", asc2 = "n"),
    panel = TRUE,
    iterlim = 2000
  )
  null_ll <- as.numeric(logLik(m_null))

  cat("Fitting Model 1 (Base + Ablation)...\n")
  m1 <- mlogit(is_choice ~ base_sim + metalign_sim + asc1 + asc2 | 0,
    data = data_m,
    rpar = c(base_sim = "n", metalign_sim = "n", asc1 = "n", asc2 = "n"),
    panel = TRUE,
    iterlim = 2000
  )

  stats1 <- calculate_model_stats(m1, null_model_ll = null_ll)

  cat("Fitting Model 2 (Base + Ablation + Main)...\n")
  m2 <- mlogit(is_choice ~ base_sim + metalign_sim + metalign_sim_main + asc1 + asc2 | 0,
    data = data_m,
    rpar = c(base_sim = "n", metalign_sim = "n", metalign_sim_main = "n", asc1 = "n", asc2 = "n"),
    panel = TRUE,
    iterlim = 2000
  )

  stats2 <- calculate_model_stats(m2, null_model_ll = null_ll)
  cat(sprintf("Model 1 BIC: %.2f, Model 2 BIC: %.2f\n", stats1$bic, stats2$bic))

  log10_bf <- calculate_log10_bf(stats1$bic, stats2$bic)
  lrt <- calculate_lrt(m1, m2)
  cat(sprintf("Log10 BF (M2/M1): %.2f, LRT p-value: %.3e\n", log10_bf, lrt$p_value))

  results$model_1 <- stats1
  results$model_2 <- stats2
  results$comparison_1_vs_2 <- list(
    log10_bayes_factor = log10_bf,
    lrt = lrt
  )
}

output_file <- sprintf("data/evals/levelso1o/%s_%s_analysis_r.json", experiment_name, backbone_name)
save_results(results, output_file)
